feat: добавить поддержку БД

RodionovRL opened the pull request:
feat: добавить базу данных
feat: обеспечить работу приложения с БД
feat: добавить свойства фильма MPA и жанр
test: добавить тестирование нового функционала

RodionovRL committed:
feat: разработать структуру БД для приложения
разработать SQL запросы для быстрого старта (создание БД, заполнение БД тестовыми данными, базовые запросы к БД)

EsiaBelos commented:
Привет, Руслан! Очень красиво сделан файл ReadMe! Все задачи ТЗ учтены. Файлы со сборкой и основными запросами составлены структурированно. Могу предложить несколько улучшений. Для файла со сборкой лучше использовать ограничение NOT NULL для нужных полей, это поможет избежать появление NPE  при работе программы, а также можно проставить ограничение количества символов для полей со строками. Работа принята! Удачи в 11 ТЗ!

RodionovRL committed:
feat: добавить базу данных
feat: обеспечить работу приложения с БД
feat: добавить свойства фильма MPA и жанр
test: добавить тестирование нового функционала

RodionovRL committed:
fix: code style

RodionovRL committed:
fix: code style 2

RodionovRL committed:
fix: code style 3

Sla-als commented on src/main/java/ru/yandex/practicum/filmorate/dao/FilmDbStorage.java:
```diff
@@ -0,0 +1,246 @@
+
+
+    @Override
+    public Film updateFilm(Film film) {
```
В текущей реализации, если нет фильма для обновления, ты выбрасываешь исключение, что не всегда является лучшей практикой. Более правильным решением было бы возвращать Optional.empty() в случае отсутствия фильма.

Sla-als commented on src/main/java/ru/yandex/practicum/filmorate/dao/UserDbStorage.java:
```diff
@@ -0,0 +1,126 @@
+        String sqlQuery = "MERGE INTO FRIENDS(USER_ID, FRIEND_ID) " +
+                "VALUES (?, ?)";
+
+        getUserById(friendId);
```
Метод addFriend извлекает пользователя с friendId из базы данных, но ничего не делает с полученным объектом. Это можно рассматривать как способ проверки существования друга, но вы можете рассмотреть возможность более явной обработки этой ситуации, например, выброс исключения, если друг не существует. Также следует обрабатывать сценарий, когда пользователь пытается добавить себя в друзья.

Sla-als replied:
Метод getUserById извлекает пользователя из базы данных, но также служит методом проверки в методах addFriend, deleteFriend и getUsersFriends. Возможно, стоит переименовать этот метод или создать новый метод, например validateUserExists, чтобы цель валидации была более очевидной

Sla-als commented on src/main/java/ru/yandex/practicum/filmorate/dao/UserDbStorage.java:
```diff
@@ -0,0 +1,126 @@
+    }
+
+    @Override
+    public boolean deleteFriend(Long id, Long exFriendId) {
```
В методе deleteFriend вы только извлекаете пользователя с id из базы данных. Также важно проверить, существует ли exFriendId в базе данных, перед тем как пытаться удалить связь.

Sla-als reviewed (commented):
Здравствуйте! Меня зовут Вячеслав, я ваш код-ревьюер на эту работу.  Ваш код в целом выглядит неплохо, но есть несколько замечаний и предложений по улучшению

RodionovRL committed:
fix: updateFilm from FilmDbStorage должен возвращать Optional
fix: выделить проверку наличия пользователя в отдельный метод checkUserIsExist
fix: при удалении из друзей проверять friendId на наличие пользователя в базе

RodionovRL committed:
fix: checkUserIsExist - применить jdbcTemplate.queryForObject  в сочетании с Optional.ofNullable

RodionovRL committed:
fix: checkUserIsExist - применить jdbcTemplate.queryForObject в конструкции try catch и отловить EmptyResultDataAccessException

Sla-als commented:
Поздравляю с успешно выполненной работой! Всё было сделано чётко и продуманно. Хотя у меня было несколько замечаний, вы справились с проектом на ура. Успехов!)
